File: swift/DeepFind.swift
// gryphon output: main.kt

import Foundation

public typealias ResultMap = [String: Any?]

extension ResultMap {
    private var arrayRegex: String {
        "\\[[0-9]+\\]$"
    }

    private func getArrayValue(_ path: String) -> Any? {
        if path.range(of: ".*\(arrayRegex)", options: .regularExpression) != nil,
            let rangeOfIndex = path.range(of: arrayRegex, options: .regularExpression)
        {
            let index = String(path[rangeOfIndex].dropFirst().dropLast())

            let pathWithoutIndex = String(path.replacingCharacters(in: rangeOfIndex, with: ""))

            let resultMap = self[pathWithoutIndex] as? [Any]

            if let intIndex = Int(index), resultMap?.indices.contains(intIndex) ?? false {
                return resultMap?[intIndex]
            }
        }

        return nil
    }

    func deepFind(_ path: String) -> Any? {
        let splittedPath = path.split(separator: ".")

        if splittedPath.count > 1 {
            if let firstPath = splittedPath.first,
                let range = Range(
                    NSRange(
                        location: firstPath.startIndex.utf16Offset(in: path),
                        length: firstPath.endIndex.utf16Offset(in: path)
                    ),
                    in: path
                )
            {
                let nextPath = String(path.replacingCharacters(in: range, with: "").dropFirst())

                if let arrayValue = getArrayValue(String(firstPath)) {
                    return (arrayValue as? ResultMap)?.deepFind(nextPath)
                }

                let resultMap = self[String(firstPath)] as? ResultMap
                return resultMap?
                    .deepFind(nextPath)
            }

            return nil
        }

        return getArrayValue(path) ?? self[path] ?? nil
    }

    func getValue(at path: String) -> Any? {
        return deepFind(path)
    }
}

File: kotlin/src/main/kotlin/hAnalytics/DeepFind.kt
typealias ResultMap = Map<String, Any?>

private val ResultMap.arrayRegex: String
    get() = "\\[[0-9]+\\]\$"

private fun ResultMap.getArrayValue(path: String): Any? {
    val rangeOfIndex: IntRange? = path.range(arrayRegex, String.CompareOptions.regularExpression)
    if (path.range(".*${arrayRegex}", String.CompareOptions.regularExpression) != null &&
            rangeOfIndex != null
    ) {
        val index: String = path[rangeOfIndex].drop(1).dropLast(1)
        val pathWithoutIndex: String = path.replacingCharacters(rangeOfIndex, "")
        val resultMap: List<Any>? = this[pathWithoutIndex] as? List<Any>
        val intIndex: Int? = index.toIntOrNull()

        if (intIndex != null && resultMap?.indices?.contains(intIndex) ?: false) {
            return resultMap?.get(intIndex)
        }
    }
    return null
}

internal fun ResultMap.deepFind(path: String): Any? {
    val splittedPath: List<String> = path.split('.')
    if (splittedPath.size > 1) {
        val firstPath: String? = splittedPath.firstOrNull()
        val range: IntRange? =
            Range(NSRange(0.utf16Offset(path), firstPath?.length?.utf16Offset(path)), path)

        if (firstPath != null && range != null) {
            val nextPath: String = path.replacingCharacters(range, "").drop(1)
            val arrayValue: Any? = getArrayValue(path = firstPath)

            if (arrayValue != null) {
                return (arrayValue as? ResultMap)?.deepFind(path = nextPath)
            }

            val resultMap: ResultMap? = this[firstPath] as? ResultMap

            return resultMap?.deepFind(path = nextPath)
        }

        return null
    }
    return getArrayValue(path = path) ?: this[path] ?: null
}

internal fun ResultMap.getValue(path: String): Any? {
    return deepFind(path = path)
}
